tests/cephfs/snapshot_clone/cephfs_snap_utils.py

```diff
@@ -0,0 +1,435 @@
+                        )
+                    ):
+                        sched_verified = 1
+                        snap_count += 1
+                if sched_type == "H|h":
+                    hour_val2 = sched_snap_list[j].split("-")[4].split("_")[0]
+                    hour_val1 = sched_snap_list[i].split("-")[4].split("_")[0]
+                    if int(sched_num) == abs(int(int(hour_val2) - int(hour_val1))):
+                        sched_verified = 1
+                        snap_count += 1
+        if sched_verified == 0:
+            log.info(out)
+            log.info(
+                f"Snapshots are NOT created as per schedule : schedule_value {sched_val} Snap Count {snap_count}"
+            )
+            return 1
+        log.info(
+            f"Snapshots are created as per schedule : schedule_value {sched_val} Snap Count {snap_count}"
+        )
+        return 0
+
+    def create_snap_retention(self, snap_params):
+        """
+        Installs all the required rpms and clones the tools required for running Tests on clients
+        Args:
+            client: ceph client to run cmd
+        Returns: None
+        """
+        client = snap_params["client"]
+        sched_cmd = f"ceph fs snap-schedule retention add {snap_params['path']} {snap_params['retention']}"
+        out, rc = client.exec_command(sudo=True, cmd=sched_cmd)
+        log.info(out)
+        if snap_params["validate"] is True:
+            cmd = f"ceph fs snap-schedule status {snap_params['path']} --fs {snap_params['fs_name']} -f json"
+            out, rc = client.exec_command(sudo=True, cmd=cmd)
+            sched_status = json.loads(out)
+            for sched_item in sched_status:
+                if sched_item["path"] == snap_params["path"] and sched_item.get(
+                    "retention"
+                ):
+                    for key, value in sched_item["retention"].items():
+                        if (f"{key} {value}" in snap_params["retention"]) or (
+                            f"{value}{key}" in snap_params["retention"]
+                        ):
+                            log.info("Snap Retention is verified.")
+                    return 0
+            log.error("Snap retention verification failed : {out}")
+        return 0
+
+    def remove_snap_retention(self, client, path, **kw_args):
+        """
+        To remove a snap-retention for given snap-schedule path
+        Args:
+        Required:
+            client: ceph client to run cmd
+            path : a snap-schedule path whose retention needs to be removed, type - str
+        Optional:
+            ret_val : a retention value that needs to be removed, for eg., "3M" or "M 3"
+        Returns: None
+        """
+        cmd = f"ceph fs snap-schedule retention remove {path}"
+        if kw_args.get("ret_val"):
+            cmd += f" {kw_args.get('ret_val')}"
+        client.exec_command(sudo=True, cmd=cmd)
+
+    def add_snap_sched_data(self, client, io_path, run_time):
```

Comment:
can't we make use of the existing IO function from cephfsutils  ?

Reply:
i need both data and metadata intensive io for each snapshot, for given duration with wait time before next iteration, so that snapshot is created. run duration is related snapshot schedule, as am running io until required snapshots are created.

And existing io modules in fsutils are,
run_ios -> it does dd or small file io, also not for given duration. new file_name for dd io and mounting_dir for smallfile io needs to be passed from test script.
write_io -> it runs for given duration but just dd.
I have planned to rework on fsutils io modules to make it generalized especially with io_params for flexbility and wide use. post that, exiting io_modules in fs_utils can be considered.